### PR TITLE
Create deadlocks in a multi-cromwell.

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

**Comment by cjllanwarne** [comment]
> Create deadlocks in a multi-cromwell

... wait, what?

**Comment by kshakir** [comment]
See #4414 for more info. We needed a reproducible test for deadlocks.

This adds PR adds a deadlock-creating-test at the high-level using multiple cromwells in a Docker Compose.

FYI- an earlier PR #4415 added very-very-low level deadlocks at the SQL/Database interface level.

File: database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala [hunk 1]
```diff
@@ -159,10 +160,30 @@ abstract class SlickDatabase(override val originalDatabaseConfig: Config) extend
...
   }
 
+  private val debugExitConfigPath = "danger.debug.only.exit-on-rollback-exception-with-status-code"
+  private val debugExitStatusCodeOption = ConfigFactory.load.getAs[Int](debugExitConfigPath)
+
   protected[this] def runTransaction[R](action: DBIO[R], isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead): Future[R] = {
-    //database.run(action.transactionally) <-- See comment above private val actionThreadPool
-    val dbio = action.transactionally.withTransactionIsolation(isolationLevel)
-    Future(Await.result(database.run(dbio), Duration.Inf))(actionExecutionContext)
+    runAutoCommit(action.transactionally.withTransactionIsolation(isolationLevel))
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala]: Discussed with faces, this seemed contradictory but is actually just a naming issue.

File: database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala [hunk 2]
```diff
@@ -159,10 +160,30 @@ abstract class SlickDatabase(override val originalDatabaseConfig: Config) extend
...
+  }
+
+  protected[this] def runAutoCommit[R](action: DBIO[R]): Future[R] = {
+    //database.run(action) <-- See comment above private val actionThreadPool
+    Future {
+      try {
+        Await.result(database.run(action), Duration.Inf)
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala]: Is this functionally different from a Future with a recover for `MySQLTransactionRollbackException`?

File: src/ci/bin/testDockerDeadlock.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+docker image ls -q broadinstitute/cromwell:"${TEST_CROMWELL_TAG}" | grep . || \
+CROMWELL_SBT_DOCKER_TAGS="${TEST_CROMWELL_TAG}" sbt server/docker
+
+# Turn off exit-on-error temporarily, as we expect an error
+set +o errexit
+CROMWELL_TAG="${TEST_CROMWELL_TAG}" \
+docker-compose \
+  -f scripts/docker-compose-mysql/docker-compose-deadlock.yml \
+  up \
+  --scale cromwell=3 \
```
Comment (salonishah11) [inline src/ci/bin/testDockerDeadlock.sh]: Just out of curiosity, why scale to only 3?

Comment (kshakir) [reply src/ci/bin/testDockerDeadlock.sh]: Shrug. 3 was what was in Thibault's original example, and worked well enough.

File: database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala (hunk 2 above)
Comment (kshakir) [reply database/sql/src/main/scala/cromwell/database/slick/SlickDatabase.scala]: Yes, it's _slightly_ different.

`try/catch` doesn't create a new `Future`, while using [`recover`](https://www.scala-lang.org/api/2.12.6/scala/concurrent/Future.html) would. `recover` requires plumbing through an execution context to create its new `Future`. That non-database EC is currently not plumbed through (but easily could be). Plus IMHO this java-esque hack is working well enough around the already existing hacks.

**Pull request merged**